pythran/pythonic/include/numpy/partial_sum.hpp

```diff
@@ -9,39 +9,44 @@ namespace pythonic
   namespace numpy
   {
 
-    template <class Op, size_t N>
-    struct _partial_sum;
-
-    template <class Op>
-    struct _partial_sum<Op, 1> {
-      template <class E, class F, class A>
+    template <class Op, size_t N, class A>
+    struct _partial_sum {
+      template <class E, class F>
+      A operator()(E e, F &o);
+      template <class E, class F>
       A operator()(E e, F &o, A acc);
     };
 
-    template <class Op, size_t N>
-    struct _partial_sum {
-      template <class E, class F, class A>
+    template <class Op, class A>
+    struct _partial_sum<Op, 1, A> {
+      template <class E, class F>
+      A operator()(E e, F &o);
+      template <class E, class F>
       A operator()(E e, F &o, A acc);
     };
 
-    template <class Op, class E,
-              class dtype = types::dtype_t<typename E::dtype>>
+    template <class Op, class E>
+    using result_dtype = types::dtype_t<decltype(std::declval<Op>()(
+        std::declval<typename std::remove_reference<E>::type::dtype>(),
+        std::declval<typename std::remove_reference<E>::type::dtype>()))>;
```

Comment:
First I though it was enough but I had some bugs. You can look at equal. Inputs may be two int arrays but the result is a bool array.